### PR TITLE
Add ElementCoordinatesMappingLocal class

### PR DESCRIPTION
**Pull request by norihiro-w**
This PR is  a part of #595 and adds mapping of lower dimensional elements (e.g. line in 2D, tri in 3D) between global and local coordinates. The mapping functionality will be used during calculation of shape functions and mapping of tensor material properties like permeability.

### CODE REVIEW

**Comment by norihiro-w** [comment]
Remark: one of the tests is not working right now due to an issue in computeRotationMatrixToXY.(). Tom will fix it.

File: GeoLib/AnalyticalGeometry.h [hunk 1]
```diff
@@ -60,11 +60,28 @@ Orientation getOrientation (const GeoLib::Point* p0,
  * @param plane_normal the normal of the plane the polygon is located in
  * @param d parameter from the plane equation
  */
-void getNewellPlane (const std::vector<GeoLib::Point*>& pnts,
+template <class T_POINT>
+void getNewellPlane (const std::vector<T_POINT*>& pnts,
```
Comment (endJunction) [inline GeoLib/AnalyticalGeometry.h]: @TomFischer Who is "Newell"?
Sarcasm aside, from the documentation it is not clear what is modified and how.

Comment (TomFischer) [inline GeoLib/AnalyticalGeometry.h]: You are right, we should rename the method to something like `getSupportingPlane`. The algorithm uses Newell's method, described in the book "Real-Time Collision Detection" from Christer Ericson. This fact is also missing in the documentation.
@norihiro-w Should I update the branch or should we make a separat PR?

Comment (norihiro-w) [inline GeoLib/AnalyticalGeometry.h]: you can make a separate PR

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        if (global_dim == 2)
+            GeoLib::compute2DRotationMatrixToX(xx, matR);
+        else
+            GeoLib::compute3DRotationMatrixToX(xx, matR);
+        matR.transposeInPlace();
+    } else if (global_dim == 3 && e.getDimension() == 2) {
+        std::vector<MathLib::Point3d*> pnts;
+        for (auto& p : vec_pt)
+            pnts.push_back(const_cast<MeshLib::Node*>(&p));
+#if 0
```
Comment (wenqing) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: What's this?

Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: these are testing another algorithm to compute a rotation matrix (actually the one used in OGS5). Once Tom fixes GeoLib::computeRotationMatrixToXY(), I'll remove them.

Comment (wenqing) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: I see.

**Comment by norihiro-w** [comment]
This PR should be merged after #659 by Tom. I will rebase my commits to it

**Comment by norihiro-w** [comment]
rebased to #659

File: MeshLib/CoordinateSystem.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+{
+
+class Element;
+
+/**
+ * \brief Coordinate system type
+ */
+struct CoordinateSystemType
+{
+    enum type
```
Comment (endJunction) [inline MeshLib/CoordinateSystem.h]: Better to use `enum class` directly.

Comment (norihiro-w) [inline MeshLib/CoordinateSystem.h]: i initially tried to use it but changed it to this way  because the code is shorter and simpler as I don't need to make cast (from enum class to int) each time whenever bit operation is used.

File: GeoLib/AnalyticalGeometry-impl.h [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    rot_mat(1,0) = -sin_theta;
+}
+
+template <class T_MATRIX>
+void compute3DRotationMatrixToX(MathLib::Vector3  const& v,
+        T_MATRIX & rot_mat)
+{
+    // a vector on the plane
+    MathLib::Vector3 yy(0, 0, 0);
+    if (fabs(v[0]) > 0.0 && fabs(v[1]) + fabs(v[2]) < std::numeric_limits<double>::epsilon()) {
```
Comment (TomFischer) [inline GeoLib/AnalyticalGeometry-impl.h]: It is also possible to use ```std::abs``` instead of ```fabs``` here and in line 55 as well as line 57 and 61.

File: GeoLib/AnalyticalGeometry-impl.h [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    yy.normalize();
+
+    for (unsigned i=0; i<3; ++i) {
+        rot_mat(0, i) = v[i];
+        rot_mat(1, i) = yy[i];
+        rot_mat(2, i) = zz[i];
+    }
+}
+
+template <class T_MATRIX>
```
Comment (TomFischer) [inline GeoLib/AnalyticalGeometry-impl.h]: I am unsure if this version of `computeRotationMatrixToXY` is still necessary?

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+namespace MeshLib
+{
+
+ElementCoordinatesMappingLocal::ElementCoordinatesMappingLocal(
+    const Element& e,
+    const CoordinateSystem &global_coords)
+: _coords(global_coords)
+{
+    assert(e.getDimension() <= global_coords.getDimension());
+    for(size_t i = 0; i < e.getNNodes(); i++)
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: `size_t`-> `std::size_t`

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    Eigen::VectorXd x_new = Eigen::VectorXd::Zero(3);
+    for(std::size_t i = 0; i < ele.getNNodes(); i++)
+    {
+        for (std::size_t j=0; j<global_dim; j++)
+            dx[j] = vec_pt[i].getCoords()[j];
+
+        x_new.head(global_dim) = matR2local * dx;
+        local_pt[i] = MeshLib::Node(x_new.data());
+    }
+};
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Semicolon is not needed.

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+void ElementCoordinatesMappingLocal::getRotationMatrixToGlobal(
+    const Element &e,
+    const CoordinateSystem &global_coords,
+    const std::vector<MeshLib::Node> &vec_pt,
+    RotationMatrix &matR) const
+{
+    const std::size_t global_dim = global_coords.getDimension();
+
+    // compute R in x=R*x' where x is original coordinates and x' is local coordinates
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: where x **are** original coordinates and x' **are** local coordinates

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    } else if (global_dim == 3 && e.getDimension() == 2) {
+        std::vector<MathLib::Point3d*> pnts;
+        for (auto& p : vec_pt)
+            pnts.push_back(const_cast<MeshLib::Node*>(&p));
+
+        // get plane normal
+        MathLib::Vector3 plane_normal;
+        double d;
+        GeoLib::getNewellPlane (pnts, plane_normal, d);
+        //std::cout << "pn=" << plane_normal << std::endl;
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Should the debug output be kept? If yes, please add a reason.

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        // compute a rotation matrix to XY
+        MathLib::DenseMatrix<double> matToXY(3,3,0.0);
+        GeoLib::computeRotationMatrixToXY(plane_normal, matToXY);
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Do you use Eigen matrices because you want to use `transposeInPlace()`? Is this the only feature you want to use from Eigen matrix implementation that is not implemented in our `DenseMatrix` class?

One possibility to avoid the copy of matrices in line 90-92 would be to make `computeRotationMatrixToXY` a template also accepting Eigen matrices.

Or we could implement `DenseMatrix<T>::transposeInPlace` and avoid using Eigen matrix class here.

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    const std::size_t global_dim = global_coords.getDimension();
+    Eigen::VectorXd dx = Eigen::VectorXd::Zero(global_dim);
+    Eigen::VectorXd x_new = Eigen::VectorXd::Zero(3);
+    for(std::size_t i = 0; i < ele.getNNodes(); i++)
+    {
+        for (std::size_t j=0; j<global_dim; j++)
+            dx[j] = vec_pt[i].getCoords()[j];
+
+        x_new.head(global_dim) = matR2local * dx;
+        local_pt[i] = MeshLib::Node(x_new.data());
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Is it possible to overload the operator* for the Eigen matrix using our `MeshLib::Node` as an argument? The it would be possible to write
```local_pt[i] = matR2local * vec_pt[i];```
instead of lines 48 - 52.

As an alternative we could use our `DenseMatrix` as an rotation matrix. Here we could implement such an operator. Maybe we can have short discussion if it is worth and if you need support.

File: MeshLib/ElementCoordinatesMappingLocal.h [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+#include "MeshLib/Elements/Element.h"
+#include "MeshLib/CoordinateSystem.h"
+
+namespace MeshLib
+{
+
+#ifdef OGS_USE_EIGEN
+typedef Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> RotationMatrix;
+#endif
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.h]: I am unsure, but I think the whole class depends on the availability of Eigen.

File: MeshLib/ElementCoordinatesMappingLocal.h [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    ElementCoordinatesMappingLocal(const Element &e, const CoordinateSystem &global_coord_system);
+	
+    /// Destructor
+    virtual ~ElementCoordinatesMappingLocal() {}
+
+    /// return the global coordinate system
+    const CoordinateSystem getGlobalCoordinateSystem() const { return _coords; }
+
+    /// return mapped coordinates of the node
+    const MeshLib::Node* getMappedCoordinates(size_t node_id) const
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.h]: `size_t` -> `std::size_t`

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+        nodes[1] = new MeshLib::Node(0.0, 0.0,  1.0);
+        return new MeshLib::Line(nodes);
+    }
+
+    static MeshLib::Line* createXY()
+    {
+        // 45degree inclined
+        MeshLib::Node** nodes = new MeshLib::Node*[e_nnodes];
+        nodes[0] = new MeshLib::Node(0.0, 0.0, 0.0);
+        nodes[1] = new MeshLib::Node(2./sqrt(2), 2./sqrt(2), 0.0);
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: 2./sqrt(2) == sqrt(2) :smile:

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    // Element information
+    typedef MeshLib::Quad ElementType;
+    static const unsigned dim = 2; //ElementType::dimension;
+    static const unsigned e_nnodes = ElementType::n_all_nodes;
+
+    // 2.5D case: inclined
+    static MeshLib::Quad* createXYZ()
+    {
+        MeshLib::Node** nodes = new MeshLib::Node*[e_nnodes];
+        // rotate 40 degree around x axis
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: 40 or 45 degree? The values looks like cos(pi/4) and sin (pi/4) and pi/4 is corresponding to 45 degree.

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+        nodes[2] = new MeshLib::Node(0.0, -1.0, -1.0);
+        nodes[3] = new MeshLib::Node(0.0,  1.0, -1.0);
+        return new MeshLib::Quad(nodes);
+    }
+
+};
+
+void debugOutput(MeshLib::Element *ele, MeshLib::ElementCoordinatesMappingLocal &mapping)
+{
+    std::cout.precision(12);
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: Why not logog? Do you plan to put this code between the preprocessor directives `#ifndef NDEBUG` and `#endif`?

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+    //debugOutput(ele, mapping);
+
+    // results when using GeoLib::ComputeRotationMatrixToXY()
+    double exp_R[3*3] = {  1, 0,  0,
+                           0, 0, -1,
+                           0, 1,  0}; //row major
+//    // results when using GeoLib::ComputeRotationMatrixToXY2()
+//    double exp_R[3*3] = { -1,  0,  0,
+//                           0,  0, -1,
+//                           0, -1,  0}; //row major
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: Is it possible to remove the commented lines?

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    auto ele = TestQuad4::createYZ();
+    MeshLib::ElementCoordinatesMappingLocal mapping(*ele, MeshLib::CoordinateSystem(*ele));
+    auto matR(mapping.getRotationMatrixToGlobal());
+    //debugOutput(ele, mapping);
+
+    // results when using GeoLib::ComputeRotationMatrixToXY()
+    double exp_R[3*3] = { 0, 0, 1,
+                          0, 1, 0,
+                         -1, 0, 0}; //row major
+//    // results when using GeoLib::ComputeRotationMatrixToXY2()
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: see above

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    //debugOutput(ele, mapping);
+
+    // results when using GeoLib::ComputeRotationMatrixToXY()
+    double exp_R[3*3] = {  1, 0, 0,
+                           0, 0.70710678118654757, -0.70710678118654757,
+                           0, 0.70710678118654757,  0.70710678118654757}; //row major
+//    // results when using GeoLib::ComputeRotationMatrixToXY2()
+//    double exp_R[3*3] = { -1, 0, 0,
+//                           0, -0.70710678118654757, -0.70710678118654757,
+//                           0, -0.70710678118654757,  0.70710678118654757}; //row major
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: see above

**Comment by TomFischer** [comment]
@norihiro-w Please let me know if I should help you fixing some of the issues.

File: MeshLib/ElementCoordinatesMappingLocal.cpp (hunk 10 above)
Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Eigen is used here because the rotation matrix will be later multiplied with shape matrices which I assume Eigen is used as data type

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp (hunk 14 above)
Comment (norihiro-w) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: yes but I intended to write it so since it's clear that the line has a length of 2.

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp (hunk 15 above)
Comment (norihiro-w) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: right, it's 45degree

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp (hunk 16 above)
Comment (norihiro-w) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: I prefer cout because logog add prefix in output. I'll add #ifdef guard

File: MeshLib/ElementCoordinatesMappingLocal.h (hunk 13 above)
Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.h]: changed it to unsigned :white_check_mark:

File: MeshLib/ElementCoordinatesMappingLocal.cpp (hunk 9 above)
Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: deleted :white_check_mark:

File: MeshLib/ElementCoordinatesMappingLocal.cpp (hunk 6 above)
Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: changed it to unsigned

File: MeshLib/ElementCoordinatesMappingLocal.cpp (hunk 11 above)
Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: technically yes as we have the code, however I would not do it for the third-party library. Instead i shortened the code using Eigen::Map

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+void ElementCoordinatesMappingLocal::rotateToLocal(
+    const Element &ele,
+    const CoordinateSystem &global_coords,
+    const RotationMatrix &matR2local,
+    std::vector<MeshLib::Node*> &vec_pt) const
+{
+    // rotate the point coordinates
+    for(unsigned i = 0; i < ele.getNNodes(); i++)
+    {
+        Eigen::Vector3d  x_new = matR2local * Eigen::Map<Eigen::Vector3d>(const_cast<double*>(vec_pt[i]->getCoords()));
```
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Is there an obstacle to implement something like
```
MeshLib::Node operator*(EigenMatrix const& mat, MeshLib::Node const& node)
```
or
```
MathLib::Point3d operator*(EigenMatrix mat, MathLib::Point3d const& )?
```
I would prefer the second version, which is more general.

This would be 
(1) more readable for the user and
(2) save some lines of code and
(3) omit some copies.

Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: updated. please see ee35769

**Comment by norihiro-w** [comment]
It appears to me that DenseMatrix is used in GeoLib and MeshLib, whereas Eigen is used in NumLib. Should I move also ElementCoordinatesMappingLocal class to NumLib since it uses Eigen?

File: Tests/MeshLib/TestCoordinatesMappingLocal.cpp [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+        nodes[1] = new MeshLib::Node(0.0, -1.0,  1.0);
+        nodes[2] = new MeshLib::Node(0.0, -1.0, -1.0);
+        nodes[3] = new MeshLib::Node(0.0,  1.0, -1.0);
+        return new MeshLib::Quad(nodes);
+    }
+
+};
+
+#ifndef NDEBUG
+void debugOutput(MeshLib::Element *ele, MeshLib::ElementCoordinatesMappingLocal &mapping)
```
Comment (TomFischer) [inline Tests/MeshLib/TestCoordinatesMappingLocal.cpp]: The function is not used and this fact gives a warning when compiling in the debug mode (tested it with clang).

File: MathLib/Point3d.h [hunk 22]
```diff
@@ -16,7 +16,12 @@
 
 #include <limits>
 
+#ifdef OGS_USE_EIGEN
+#include <Eigen/Eigen>
+#endif
```
Comment (endJunction) [inline MathLib/Point3d.h]: @TomFischer Is eigen needed now?

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+{
+	for (auto p : _point_vec) delete p;
+}
+
+void ElementCoordinatesMappingLocal::rotateToLocal(
+    const RotationMatrix &matR2local,
+    std::vector<MeshLib::Node*> &vec_pt) const
+{
+    for(unsigned i = 0; i < vec_pt.size(); i++)
+        vec_pt[i]->setCoords((matR2local* (*vec_pt[i])).getCoords());
```
Comment (endJunction) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: More style than performance. If I'm not wrong this could be
```c++
for(MeshLib::Node* node : vec_pt)
    node->setCoords((matR2local* (*node)).getCoords());
```
Also there is a mismatch between vector of Node (pointers) and its name vec_pt.

File: MeshLib/ElementCoordinatesMappingLocal.cpp [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+void ElementCoordinatesMappingLocal::getRotationMatrixToGlobal(
+    const Element &e,
+    const CoordinateSystem &global_coords,
+    const std::vector<MeshLib::Node*> &vec_pt,
```
Comment (endJunction) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: Same naming mismatch here. Maybe `std::vector<MathLib::Point3d>` is sufficient here; No `Node` features seem to be used.

File: MeshLib/ElementCoordinatesMappingLocal.h [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+
+    /// get a rotation matrix to the global coordinates
+    /// it computes R in x=R*x' where x is original coordinates and x' is local coordinates
+    void getRotationMatrixToGlobal(
+            const Element &e, const CoordinateSystem &coordinate_system,
+            const std::vector<MeshLib::Node*> &vec_pt, RotationMatrix &matR2original) const;
+
+private:
+    const CoordinateSystem _coords;
+    std::vector<MeshLib::Node*> _point_vec;
```
Comment (endJunction) [inline MeshLib/ElementCoordinatesMappingLocal.h]: Again, `point_vec` is a vector of nodes...

File: MathLib/Point3d.h (hunk 22 above)
Comment (TomFischer) [inline MathLib/Point3d.h]: I think Eigen is not needed here. I forgot to remove these lines. @norihiro-w : can you remove the lines and merge the changes to the corresponding commit. By the way, I think the commits  c7deda4 and  ee35769 can also be squashed together. If you want do this go ahead.

File: MeshLib/ElementCoordinatesMappingLocal.cpp (hunk 24 above)
Comment (TomFischer) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: The original "point" data are the `MeshLib::Node`s from an element. At least pointers to `Point3d` are needed for polymorphism.

File: MathLib/Point3d.h (hunk 22 above)
Comment (norihiro-w) [inline MathLib/Point3d.h]: deleted and squashed

File: MeshLib/ElementCoordinatesMappingLocal.cpp (hunk 24 above)
Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.cpp]: renamed to vec_nodes

File: MeshLib/ElementCoordinatesMappingLocal.h [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+
+#include "MeshLib/Elements/Element.h"
+#include "MeshLib/CoordinateSystem.h"
+
+namespace MeshLib
+{
+
+#ifdef OGS_USE_EIGEN
+typedef Eigen::Matrix<double, 3u, 3u, Eigen::RowMajor> RotationMatrix;
+#endif
```
Comment (endJunction) [inline MeshLib/ElementCoordinatesMappingLocal.h]: If eigen is not found or is disabled, the type RotationMatrix is not defined causing a compilation error. We need a default value.

@TomFischer Which parts of the interface are missing in DenseMatrix for this code to compile?
I found only `RotationMatrix::Identity()` to be missing...

Comment (norihiro-w) [inline MeshLib/ElementCoordinatesMappingLocal.h]: added DenseMatrix as a default case

**Comment by norihiro-w** [comment]
tests are failing on Windows MSVC while it is fine on GCC and Clang. as i dont have windows environment at home, i will check this after i am back in the office.

**Comment by bilke** [comment]
Jenkins test this please!

**Comment by norihiro-w** [comment]
got the following errors with MSVC. is there anyone having idea what to do?

```
Error	28	error C2280: 'ProcessLib::GroundwaterFlow::LocalAssemblerData<NumLib::ShapeLine2,NumLib::IntegrationGaussRegular<0x01>,GlobalMatrix_,GlobalVector_>::LocalAssemblerData(void)' : attempting to reference a deleted function	C:\Users\watanabe\Documents\GitHub\ogs\AssemblerLib\LocalDataInitializer.h	59	1	ogs
Error	35	error C2542: 'NumLibFemIsoTest<Type>::D' : class object has no constructor for initialization	C:\Users\watanabe\Documents\GitHub\ogs\Tests\NumLib\TestFe.cpp	104	1	testrunner
Error	36	error C2542: 'NumLibFemIsoTest<Type>::expectedM' : class object has no constructor for initialization	C:\Users\watanabe\Documents\GitHub\ogs\Tests\NumLib\TestFe.cpp	104	1	testrunner
Error	37	error C2542: 'NumLibFemIsoTest<Type>::expectedK' : class object has no constructor for initialization	C:\Users\watanabe\Documents\GitHub\ogs\Tests\NumLib\TestFe.cpp	104	1	testrunner
Error	38	error C1903: unable to recover from previous error(s); stopping compilation	C:\Users\watanabe\Documents\GitHub\ogs\Tests\NumLib\TestFe.cpp	109	1	testrunner
```

**Comment by endJunction** [comment]
I compiled the code with clang in RelWithDebInfo, which run through OK. But running tests ended in a SEGV for the:
```
[----------] 3 tests from NumLibFemIsoTest/5, where TypeParam = (anonymous namespace)::TestCase<FeTestData::TestFeLINE2, EigenFixedShapeMatrixPolicy>
[ RUN      ] NumLibFemIsoTest/5.CheckMassMatrix
```
test.
Running it in gdb does not always result in a SEGV: output and the backtrace [here](https://gist.github.com/endJunction/6473dbb722eda3f3c509)

Valgrind's memcheck does not show any invalid reads/writes until a SEGV. (see the above gist for details)

**Comment by norihiro-w** [comment]
update: the test failing on Jenkins cannot be reproduced on my windows (win7, MSVC2013 with update 4, 64bit). could be there is something unitialized or machine dependent bug.

**Comment by norihiro-w** [comment]
rebased to HEAD

**Comment by norihiro-w** [comment]
@bilke Eigen version used in OGS6/Win-PRs seems to be bit old (3.1.4). Can you please replace it with the latest (3.2.4)?

**Comment by bilke** [comment]
I updated eigen to 3.2.4.

Jenkins test this please!

**Comment by bilke** [comment]
test this please

**Comment by norihiro-w** [comment]
@bilke thanks!

**Comment by norihiro-w** [comment]
the problem was solved!
